cuda/trsv_kernels.cuh

```diff
@@ -0,0 +1,1035 @@
+            (dmtx == dmtx_t::unit && col == row)
+                ? ValueType{1}
+                : (row <= col && 0 <= global_row && 0 <= global_col)
+                      ? mtx[global_row * m_info.stride + global_col]
+                      : ValueType{0};
+    }
+    group.sync();
+    // Invert lower triangular matrix
+
+    // First: make diagonals 1
+    if (dmtx == dmtx_t::non_unit) {
+        if (threadIdx.y == 0) {
+            for (int row = swarp_size; row >= 0; --row) {
+                const int col = threadIdx.x;
+                const auto triang_idx = col * triang_stride + row;
+                const auto local_val = triang[triang_idx];
+                const auto diag_el = ValueType{1} / swarp.shfl(local_val, row);
+                if (row <= col) {
+                    triang[triang_idx] =
+                        (row == col) ? diag_el : local_val * diag_el;
+                }
+            }
+        }
+    }
+    // No need to sync since the same warp will continue to compute
+
+    if (threadIdx.y == 0) {
+        // Compute remaining inverse in-place (with Gauss-Jordan elimination)
+        for (int diag_id = swarp_size - 1; diag_id >= 0; --diag_id) {
+            const int col = threadIdx.x;
+            const auto diag_el = triang[diag_id * triang_stride + diag_id];
+            for (int row = diag_id - 1; row >= 0; --row) {
+                // broadcast
+                const auto factor = -triang[diag_id * triang_stride + row];
+                const int triang_idx = col * triang_stride + row;
+                const auto triang_val = triang[triang_idx];
+                if (row < col) {
+                    triang[triang_idx] =
+                        (col == diag_id)
+                            ? factor * diag_el
+                            : triang_val +
+                                  factor *
+                                      triang[col * triang_stride + diag_id];
+                }
+                swarp.sync();
+            }
+        }
+    }
+    group.sync();
+
+    constexpr index_type num_local_rows = swarp_size / swarps_per_block;
+    ValueType local_row_result[num_local_rows] = {};
+    for (index_type col_block = 1; col_block <= row_block_idx; ++col_block) {
+        const index_type global_col =
+            m_info.size[1] - (col_block * swarp_size) + threadIdx.x;
+
+        // Wait until result is known for current column block
+        if (threadIdx.x == 0 && threadIdx.y == 0) {
+            // Note: this needs to be signed since the initial value is
+            //       ~0 (all ones)
+            volatile auto *last_col = reinterpret_cast<std::make_signed_t<
+                std::remove_pointer_t<decltype(col_row_global_helper)>> *>(
+                col_row_global_helper);
+            while (*last_col < col_block - 1) {
+            }
+        }
+        group.sync();
+        // Make sure that data is only read after synchronization
+        __threadfence();
```

Comment:
I need to look into it in more detail. it ensures all blocks see the same global memory result of writing before it, right?